Comments apply globally, affecting other files in strange ways

This might be a newb question, but I'm currently in the process of migrating JSHint -> ESLint and one thing which caught me off-guard is that the comments to change a rule do not apply on a block level or even file level like they do in JSHint, but globally, ...or something.

Here's a reduced example to show the ~~bug~~ unexpected behaviour. No custom rules in play here, no .eslintrc files in the ancestry, just eslint 0.5.0 on Node 0.10.25.

Directory contains two files: `foo.js` and `bar.js`.

``` javascript
// foo.js
function foo() { return 1; }

// bar.js
function bar() { return 2; }
```

Running `eslint` in that directory gives this output:

```
bar.js
  1:0  error  Missing "use strict" statement  strict

foo.js
  1:0  error  Missing "use strict" statement  strict
```

...as expected.

But when I change `bar.js` to this code...

``` javascript
/*eslint strict: 0*/
function bar() { return 2; }
```

I get no warnings (exit code 0) from eslint.

What I expected was that it would still error on the missing 'use strict' in `foo.js`, since that doesn't include it.

Stranger still is if, instead of changing `bar.js`, `foo.js` is changed to include `/*eslint strict: 0 */`, then running eslint just gives a _warning_ for missing 'use-strict' in `bar.js`.

I assume this would have something to do with the order in which the files are checked.

---

Is there a way to temporarily (meaning, scoped by lines, scope or file) to modify the rules? As an example of how it might work, jslint and jshint allow code like this:

``` javascript
/*jslint bitwise: false */
x = y & 3; // no error
/*jslint bitwise: true */
z = w & 3; // error
```

--- a/lib/eslint.js
+++ b/lib/eslint.js
@@ -213,6 +213,23 @@ function modifyConfigFromComments(ast, config) {
     return util.mergeConfigs(config, commentConfig);
 }
 
+/**
+ * Process initial config to make it safe to extend by file comment config
+ * @param  {Object} config Initial config
+ * @returns {Object}        Processed config
+ */
+function prepareConfig(config) {
+
+    config.globals = config.globals || config.global || {};
+    delete config.global;
+
+    return {
+        rules: util.mergeConfigs({}, config.rules || {}),
+        globals: util.mergeConfigs({}, config.globals),
+        env: util.mergeConfigs({}, config.env || {})
+    };
+}
+
 //------------------------------------------------------------------------------
 // Public Interface
 //------------------------------------------------------------------------------
@@ -310,9 +327,8 @@ module.exports = (function() {
 
         //if Esprima failed to parse the file, there's no sense in setting up rules
         if (ast) {
-
-            config.globals = config.globals || config.global || {};
-            delete config.global;
+            // process initial config to make it safe to extend
+            config = prepareConfig(config);
 
             // parse global comments and modify config
             config = modifyConfigFromComments(ast, config);

--- a/tests/lib/eslint.js
+++ b/tests/lib/eslint.js
@@ -1253,6 +1253,14 @@ describe("eslint", function() {
 
             assert.equal(messages.length, 0);
         });
+
+        it("should process empty config", function() {
+            var config = {};
+
+            eslint.reset();
+            var messages = eslint.verify(code, config, filename, true);
+            assert.equal(messages.length, 0);
+        });
     });
 
     describe("after calling reset()", function() {
@@ -1543,9 +1551,9 @@ describe("eslint", function() {
     });
 
     describe("when evaluating code with comments to enable rules", function() {
-        var code = "/*eslint no-alert:1*/ alert('test');";
 
         it("should report a violation", function() {
+            var code = "/*eslint no-alert:1*/ alert('test');";
             var config = { rules: {} };
 
             var messages = eslint.verify(code, config, filename);
@@ -1553,6 +1561,19 @@ describe("eslint", function() {
             assert.equal(messages[0].ruleId, "no-alert");
             assert.equal(messages[0].message, "Unexpected alert.");
             assert.include(messages[0].node.type, "CallExpression");
+        });
+
+        it("rules should not change inital config", function() {
+            var config = { rules: {"strict": 2} };
+            var codeA = "/*eslint strict: 0*/ function bar() { return 2; }";
+            var codeB = "function foo() { return 1; }";
+
+            eslint.reset();
+            var messages = eslint.verify(codeA, config, filename, false);
+            assert.equal(messages.length, 0);
+
+            messages = eslint.verify(codeB, config, filename, false);
+            assert.equal(messages.length, 1);
         });
     });
 

